measure_optimize.py

```diff
@@ -2,180 +2,2412 @@
+                        Sigma[sigma_i_start + t1, sigma_i_start + t2] = error_cov[i][
+                            t1
+                        ][t2]
+
+            if self.precompute_print_level >= 2:
+                print("Error covariance matrix option:", error_option)
+                if self.precompute_print_level == 3:
+                    print("Error matrix:", Sigma)
+
+        # covariance between measurements
+        # the covariances between measurements at the same timestep in a time-invariant way
+        elif error_option == CovarianceStructure.measure_correlation:
+            # loop over number of measurements
+            for i in range(self.sens_info.total_measure_idx):
+                # loop over number of measurements
+                for j in range(self.sens_info.total_measure_idx):
+                    # find the covariance term
+                    cov_ij = error_cov[i][j]
+                    # find the starting index for each measurement (each measurement i has Nt entries)
+                    head_i = self.head_pos_flatten[i]
+                    # starting index for measurement j
+                    head_j = self.head_pos_flatten[j]
+                    # i, j may have different timesteps
+                    # we find the corresponding index by locating the starting indices
+                    for t in range(self.sens_info.Nt):
+                        Sigma[t + head_i, t + head_j] = cov_ij
+
+            if self.precompute_print_level >= 2:
+                print("Error covariance matrix option:", error_option)
+                if self.precompute_print_level == 3:
+                    print("Error matrix:", Sigma)
+
+        # the full covariance matrix is given
+        elif error_option == CovarianceStructure.time_measure_correlation:
+            Sigma = np.asarray(error_cov)
+
+            if self.precompute_print_level >= 2:
+                print("Error covariance matrix option:", error_option)
+                if self.precompute_print_level == 3:
+                    print("Error matrix:", Sigma)
+
+        self.Sigma = Sigma
+
+        return Sigma
+
+    def _split_sigma(self, Sigma):
+        """Split the error covariance matrix to be used for computation
+        They are split to DCM-DCM (scalar) covariance, DCM-SCM (vector) covariance, SCCM-SCM (matrix) covariance
+        We inverse the Sigma for the computation of FIM
+
+        Arguments
+        ---------
+        :param Sigma: a 2D numpy array, covariance matrix for the flattened measurements
+            Shape: sum(Nt) * sum(Nt)
+
+        Returns
+        -------
+        None
+        """
+        # Inverse of covariance matrix is used
+        # pinv is used to avoid ill-conditioning issues
+        Sigma_inv = np.linalg.pinv(Sigma)
+        self.Sigma_inv_matrix = Sigma_inv
+        # Use a dicionary to store the inverse of sigma as either scalar number, vector, or matrix
+        self.Sigma_inv = {}
```

Comment:
Whenever declaring an empty dictionary, you need comments explaining what the values and keys will be.